pkg/csi/service/vanilla/controller.go

```diff
@@ -255,6 +267,44 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 		log.Error(msg)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
+
+	if c.manager.CnsConfig.FeatureStates.CSIMigration && scParams.CSIMigration == "true" {
+		if len(scParams.Datastore) != 0 {
+			log.Infof("Converting datastore name: %q to Datastore URL", scParams.Datastore)
+			vcList := c.manager.VcenterManager.GetAllVirtualCenters()
+			dcList, err := vcList[0].GetDatacenters(ctx)
+			if err != nil {
+				msg := fmt.Sprintf("failed to get datacenter list. err: %+v", err)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			foundDatastoreURL := false
+			for _, dc := range dcList {
+				dsURLTodsInfoMap, err := dc.GetAllDatastores(ctx)
+				if err != nil {
+					msg := fmt.Sprintf("failed to get dsURLTodsInfoMap. err: %+v", err)
+					log.Error(msg)
+					return nil, status.Errorf(codes.Internal, msg)
+				}
+				for dsURL, dsInfo := range dsURLTodsInfoMap {
+					if dsInfo.Info.Name == scParams.Datastore {
+						scParams.DatastoreURL = dsURL
```

Comment:
Retrieving the datastoreUrl for a datastore name is quite expensive in this code block as this block is executed for every create volume callback. Why not just get the datastore object from datastore name? Can we not use finder.Datastore method in govmomi?

Reply:
finder.datastore operates of datastore path and not name - https://github.com/vmware/govmomi/blob/master/find/finder.go#L445

also if multiple instances retrieved it returns an error, in that case, another call may be required. - https://github.com/vmware/govmomi/blob/master/find/finder.go#L451-L453

let me see how I can enhance this.

Reply:
Created an issue for this. Let me address this as part of next PR - https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/254